### PR TITLE
User remember me functionality 7

### PR DESCRIPTION
**Pull request by nimish13**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  def new
+  end
+
+  def create
+    @user = User.find_by(email: params[:user][:email])
+
+    if @user.try(:authenticate, params[:user][:password])
+      if @user.confirmed_at.nil?
+        redirect_to login_url, alert: 'Account not confirmed' and return
```
Comment (nimish13) [inline app/controllers/sessions_controller.rb]: change message to 'Please confirm your account to log in.'

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  def destroy
+    delete_remember_me_cookie
+    session.clear
+    redirect_to login_url, notice: 'Successfully logged out.'
+  end
+
+  private
+
+    def create_remember_me_cookie
+      # deleteing old cookie
```
Comment (nimish13) [inline app/controllers/sessions_controller.rb]: change the comment text

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      cookies.encrypted[:remember_me] = {
+        value: @user.api_token,
+        expires: 3.months.from_now,
```
Comment (nimish13) [inline app/controllers/sessions_controller.rb]: take time in a constant

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    session.clear
+    redirect_to login_url, notice: 'Successfully logged out.'
+  end
+
+  private
+
+    def create_remember_me_cookie
+      # deleteing old cookie
+      cookies.encrypted[:remember_me] = {
+        value: @user.api_token,
```
Comment (nimish13) [inline app/controllers/sessions_controller.rb]: create a new field in user.rb for `remember_me` token

File: app/mailers/application_mailer.rb [hunk 5]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: 'avneetmalhotra02@gmail.com'
```
Comment (nimish13) [inline app/mailers/application_mailer.rb]: extract this in application.yml

File: app/controllers/registrations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
+class RegistrationsController < ApplicationController
+  helper ErrorHelper
```
Comment (nimish13) [inline app/controllers/registrations_controller.rb]: is it needed?

File: app/mailers/registration_mailer.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class RegistrationMailer < ApplicationMailer
```
Comment (nimish13) [inline app/mailers/registration_mailer.rb]: Now we will be naming this mail to user_mailer

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+class User < ApplicationRecord
+  has_secure_password
+  has_secure_token :confirmation_token
+  has_secure_token :api_token
+  has_secure_token :password_reset_token
+
+  ## VALIDATIONS
+  validates :name, presence: true
```
Comment (nimish13) [inline app/models/user.rb]: use `with_options` for better readability

File: app/views/confirmations/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<div class='jumbotron confirmation-form'>
+  <%= form_with(url: confirmations_path, local: true ) do |form| %>
+    <fieldset>
+      <div class="form-group">
+        <%= form.label :email, 'Email' %>
+        <%= form.email_field :email, id: :user_email, class: 'form-control', placeholder: 'myemail@gmail.com' %>
+      </div>
+
+      <%= form.submit 'Resend Confirmation Email', class: 'btn btn-primary' %>
```
Comment (nimish13) [inline app/views/confirmations/_form.html.erb]: Change name to `Send`

File: app/views/confirmations/_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    </fieldset>
+  <% end %>
+</div>
```
Comment (nimish13) [inline app/views/confirmations/_form.html.erb]: please provide EOL

File: app/views/layouts/_flash_messages.html.erb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<% if flash.any? %>
```
Comment (nimish13) [inline app/views/layouts/_flash_messages.html.erb]: move this file to shared folder

File: app/views/layouts/_navbar.html.erb [hunk 12]
```diff
@@ -0,0 +1,46 @@
+<nav class="navbar navbar-expand-lg navbar-dark bg-primary">
```
Comment (nimish13) [inline app/views/layouts/_navbar.html.erb]: move this file to shared folder

File: app/views/registrations/_form.html.erb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        <%= display_error(:password, user) %>
+      </div>
+
+      <div class='form-group'>
+        <%= form.label :password_confirmation, 'Confirm Password', class: 'required-field' %>
+        <%= form.password_field :password_confirmation, id: :user_password_confirmation, class: 'form-control', placeholder: '******' %>
+        <%= display_error(:password_confirmation, user) %>
+      </div>
+
+      <% if user.id %>
```
Comment (nimish13) [inline app/views/registrations/_form.html.erb]: please use this partial to render only common fields

File: config/routes.rb [hunk 14]
```diff
@@ -2,4 +2,30 @@
...
   root 'home#index'
+  # controller :registrations do
+  #   get 'signup', to: 'new'
+  #   post 'signup', to: 'create'
+  # end
+
+  resources :registrations, only: [:new, :create, :edit, :update]
+
+  resources :confirmations, only: [:new, :create]
+  get 'account-confirmed', to: 'confirmations#confirm'
```
Comment (nimish13) [inline config/routes.rb]: /users/confirm

File: db/migrate/20171215061515_create_users.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
+class CreateUsers < ActiveRecord::Migration[5.1]
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :email
```
Comment (nimish13) [inline db/migrate/20171215061515_create_users.rb]: uniqueness index

File: app/controllers/registrations_controller.rb [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    def update_user_params
+      params.require(:user).permit(:name, :password, :password_confirmation)
+    end
+
+    def ensure_logged_out
+      redirect_to root_url if current_user
+    end
+
+    def set_user
+      @user = User.find_by(id: params[:id])
```
Comment (nimish13) [inline app/controllers/registrations_controller.rb]: fetch only current_user here

File: app/controllers/sessions_controller.rb [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+      session[:user_id] = @user.id
+      redirect_to root_url
+    else
+      redirect_to login_url, alert: 'Invalid email/passord. Please try again.'
+    end
+  end
+
+  def destroy
+    delete_remember_me_cookie
+    session.clear
```
Comment (nimish13) [inline app/controllers/sessions_controller.rb]: use reset_session

File: app/controllers/sessions_controller.rb [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+    if @user.try(:authenticate, params[:user][:password])
+      if @user.confirmed_at.nil?
+        redirect_to login_url, alert: 'Please confirm your account to log in.' and return
+      end
+
+      create_remember_me_cookie if params[:remember_me].present?
+
+      session[:user_id] = @user.id
+      redirect_to root_url
```
Comment (nimish13) [inline app/controllers/sessions_controller.rb]: add flash

File: app/controllers/sessions_controller.rb [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    session.clear
+    redirect_to login_url, notice: 'Successfully logged out.'
+  end
+
+  private
+
+    def login_if_remember_user
+      if cookies[:remember_me].present?
+        user = User.find_by(remember_me_token: cookies.encrypted[:remember_me])
+        session[:user_id] = user.id
```
Comment (nimish13) [inline app/controllers/sessions_controller.rb]: ensure user present and already confirmed

File: app/helpers/error_helper.rb [hunk 20]
```diff
@@ -0,0 +1,7 @@
+module ErrorHelper
+  def display_error(attribute, model_object)
+    unless model_object.errors[attribute].empty?
+      content_tag :div, attribute.to_s.capitalize + ' ' +  model_object.errors[attribute].join(), class: 'invalid-feedback display-block'
```
Comment (nimish13) [inline app/helpers/error_helper.rb]: remove this `attribute.to_s.capitalize`

File: app/mailers/user_mailer.rb [hunk 21]
```diff
@@ -0,0 +1,10 @@
+class UserMailer < ApplicationMailer
+
+  def confirmation(user)
+    @user = user
+
+    mail(to: @user.email, subject: 'Account Confimration')
+
+    @user.update(confirmation_token_sent_at: Time.current)
```
Comment (nimish13) [inline app/mailers/user_mailer.rb]: this should be set when confirmation_token is genereated

Comment (nimish13) [reply app/mailers/user_mailer.rb]: secondly when the account is confirmed the confirmation_token should be set to nil

File: app/models/user.rb [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+  end
+  validates :email, allow_blank: true, uniqueness: { case_sensitive: false }
+  validates :email, format:{
+    with: Regexp.new(ENV['email_regex']),
+    allow_blank: true
+  }
+  validates :password, allow_blank: true, length: { minimum: 6 }
+
+    def send_confrimation_instructions
+      UserMailer.confirmation(self).deliver_now
```
Comment (nimish13) [inline app/models/user.rb]: why are we not using deliver_later

Comment (nimish13) [reply app/models/user.rb]: dont pass the entire object instead pass user_id and in mailer fetch the user by id

File: app/models/user.rb [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+  has_secure_password
+  has_secure_token :confirmation_token
+  has_secure_token :api_token
+  has_secure_token :password_reset_token
+  has_secure_token :remember_me_token
+
+  ## VALIDATIONS
+  with_options presence: true do
+    validates :name, :email
+  end
```
Comment (nimish13) [inline app/models/user.rb]: can we optimize this?

File: config/environments/development.rb [hunk 24]
```diff
@@ -26,8 +26,21 @@
     config.cache_store = :null_store
   end
 
+  config.action_mailer.perform_deliveries = true
+  config.action_mailer.delivery_method = :letter_opener
+  config.action_mailer.default_url_options = { host: 'http://localhost:3000' }
+  # config.action_mailer.asset_host = 'http://localhost:3000'
+  config.action_mailer.smtp_settings = {
```
Comment (nimish13) [inline config/environments/development.rb]: is it needed when we are using letter_opener

File: config/routes.rb [hunk 25]
```diff
@@ -2,4 +2,30 @@
...
+  # resources :sessions, only: [:new, :create]
```
Comment (nimish13) [inline config/routes.rb]: remove commented code

**Comment by avneetmalhotra** [comment]
closed in reference to #11

**Pull request closed without merging**